src/VecSim/algorithms/brute_force/bf_batch_iterator.h

```diff
@@ -39,3 +41,167 @@ class BF_BatchIterator : public VecSimBatchIterator {
 
     ~BF_BatchIterator() override = default;
 };
+
+/******************** Implementation **************/
+
+// heuristics: decide if using heap or select search, based on the ratio between the
+// number of remaining results and the index size.
+template <typename DataType, typename DistType>
+VecSimQueryResult_List
+BF_BatchIterator<DataType, DistType>::searchByHeuristics(size_t n_res,
+                                                         VecSimQueryResult_Order order) {
+    if ((this->index->indexLabelCount() - this->getResultsCount()) / 1000 > n_res) {
+        // Heap based search always returns the results ordered by score
+        return this->heapBasedSearch(n_res);
+    }
+    VecSimQueryResult_List rl = this->selectBasedSearch(n_res);
+    if (order == BY_SCORE) {
+        sort_results_by_score(rl);
+    }
+    return rl;
+}
+
+template <typename DataType, typename DistType>
+void BF_BatchIterator<DataType, DistType>::swapScores(
+    const vecsim_stl::unordered_map<size_t, size_t> &TopCandidatesIndices, size_t res_num) {
```

Comment:
shouldn't it be `const vecsim_stl::unordered_map<labelType, size_t>`
@GuyAv46?

Reply:
Don't know why git blames me here, but looks like it should